File: Protogen/loxim_protocol/java/src/main/java/pl/edu/mimuw/loxim/protocol/packages_data/DatetimetzPackage.java
package pl.edu.mimuw.loxim.protocol.packages_data;

import java.util.EnumSet;
import pl.edu.mimuw.loxim.protogen.lang.java.template.ptools.Package;
import pl.edu.mimuw.loxim.protogen.lang.java.template.exception.ProtocolException;
import pl.edu.mimuw.loxim.protogen.lang.java.template.streams.PackageInputStreamReader;
import pl.edu.mimuw.loxim.protogen.lang.java.template.streams.PackageOutputStreamWriter;


import pl.edu.mimuw.loxim.protocol.enums.*;


public  class DatetimetzPackage extends Package {
		public final static long ID=(long)14;
	
    	/**
**/
    	private DatePackage date=null;
    	/**
**/
    	private TimePackage time=null;
    	/**
**/
    	private byte tz=0;
    	
	public DatetimetzPackage()
	{
	}

		public DatetimetzPackage(
		    	DatePackage a_date,
	        	TimePackage a_time,
	        	byte a_tz    	)
	{
        	date=a_date;
        	time=a_time;
        	tz=a_tz;
    
	}
		
		@Override
	public long getPackageType() {
		return (long)ID;
	}
		
	@Override
	protected void deserializeW(PackageInputStreamReader reader)
			throws ProtocolException {
		super.deserializeW(reader);
		
				try{
	    			date=new DatePackage();
date.deserializeContent(reader);
					time=new TimePackage();
time.deserializeContent(reader);
					tz=reader.readInt8();
				}catch(Exception e)
		{
			throw new ProtocolException(e);
		}
			}
	
	@Override
	protected void serializeW(PackageOutputStreamWriter writer)
			throws ProtocolException {
		super.serializeW(writer);
				try{
	    			date.serializeContent(writer);
		
					time.serializeContent(writer);
		
					writer.writeInt8(tz);		
				}catch(Exception e)
		{
			throw new ProtocolException(e);
		}	
			}
	
	@Override
	public boolean equals(Object obj) {
		if(!super.equals(obj))
			return false;
		DatetimetzPackage p=(DatetimetzPackage)obj;
	    if (!(((date==p.getDate())||((date!=null)&&(p.getDate()!=null)&&((date).equals(p.getDate())))))) 
return false;
		if (!(((time==p.getTime())||((time!=null)&&(p.getTime()!=null)&&((time).equals(p.getTime())))))) 
return false;
		if (!((tz==p.getTz()))) 
return false;
				return true;
	}
	
//================= GETTERS AND SETTERS ===================

        	public DatePackage getDate(){return date;};
		
    	public void setDate(DatePackage a_date){ date=a_date;};
		
        	public TimePackage getTime(){return time;};
		
    	public void setTime(TimePackage a_time){ time=a_time;};
		
        	public byte getTz(){return tz;};
		
    	public void setTz(byte a_tz){ tz=a_tz;};
		
    
	 
}

File: Protogen/loxim_protocol/cpp/protocol/packages_data/DatetimetzPackage.cpp
#include <string.h>
#include <stdlib.h>
#include "DatetimetzPackage.h"
#include "ptools/PackageBufferWriter.h"
#include "ptools/PackageBufferReader.h"

using namespace protocol;

DatetimetzPackage::DatetimetzPackage(\
		DatePackage* a_date,\
		TimePackage* a_time,\
		int8_t a_tz
		)
{
	date=a_date;
	time=a_time;
	tz=a_tz;
}

DatetimetzPackage::DatetimetzPackage()
{
	date=NULL;
	time=NULL;
	tz=0;
}

DatetimetzPackage::~DatetimetzPackage()
{
	if(date) delete date;
	date=NULL;
	if(time) delete time;
	time=NULL;
}

void DatetimetzPackage::serializeW(PackageBufferWriter *writer)
{
	Package::serializeW(writer);
	date->serializeContent(writer);
	time->serializeContent(writer);
	writer->writeInt8(tz);
}

bool DatetimetzPackage::deserializeR(PackageBufferReader *reader)
{
	if (!Package::deserializeR(reader)) return false;
	date=new DatePackage();
	if (!date->deserializeContent(reader)) return false;
	time=new TimePackage();
	if (!time->deserializeContent(reader)) return false;
	tz=reader->readInt8();
	return true;
}

bool DatetimetzPackage::equals(Package *_package)
{
	if (!_package) return false;
	if (_package->getPackageType()!=this->getPackageType()) return false;
	if (!Package::equals(_package)) return false;
	DatetimetzPackage* p=(DatetimetzPackage*)_package;
	if (!((!getDate() && ! p->getDate())||(getDate() && p->getDate() && (getDate()->equals(p->getDate()))))) return false;
	if (!((!getTime() && ! p->getTime())||(getTime() && p->getTime() && (getTime()->equals(p->getTime()))))) return false;
	if (!(getTz()==p->getTz())) return false;
	return true;
}
